### PR TITLE
Shoping Cart 

### PR DESCRIPTION
**Pull request by VKhmil**
Added Entities ShoppingCart, ItemCart, repos DTO layer.

### CODE REVIEW

File: src/main/java/com/bookappstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/cart")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/ShoppingCartController.java]: Use server.servlet.context-path=/api
```suggestion
@RequestMapping("/cart")
```

File: src/main/java/com/bookappstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Retrieve user's shopping cart",
+            description = "Retrieves the user's shopping cart")
+    public ShoppingCartResponseDto getShoppingCart(
+            Authentication authentication,
+            @SortDefault.SortDefaults({
+                    @SortDefault(sort = "id", direction = Sort.Direction.ASC)})
+                    Pageable pageable) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.getShoppingCartDtoByUserId(user.getId());
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/ShoppingCartController.java]:
```suggestion
        return shoppingCartService.getShoppingCart(user.getId());
```

File: src/main/java/com/bookappstore/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Add book to the shopping cart",
+            description = "Adds a book to the user's shopping cart")
+    public ShoppingCartResponseDto addCartItem(
+            Authentication authentication,
+            @RequestBody CartItemRequestDto cartItemRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/ShoppingCartController.java]:
```suggestion
            @RequestBody @Valid CartItemRequestDto cartItemRequestDto) {
```

File: src/main/java/com/bookappstore/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.addCartItemByUserId(user.getId(), cartItemRequestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/ShoppingCartController.java]:
```suggestion
        return shoppingCartService.addCartItemToCart(user.getId(), cartItemRequestDto);
```

File: src/main/java/com/bookappstore/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @PutMapping("/cart-items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Update quantity of a book in the shopping cart",
+            description = "Updates the quantity of a book in the user's shopping cart")
+    public ShoppingCartResponseDto updateCartItem(
+            Authentication authentication,
+            @PathVariable Long cartItemId,
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/ShoppingCartController.java]:
```suggestion
            @PathVariable @Positive Long cartItemId,
```
Don't forget to add `@Validated`on the controller

File: src/main/java/com/bookappstore/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            @RequestBody CartItemUpdateDto cartItemRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/ShoppingCartController.java]:
```suggestion
            @RequestBody @Valid CartItemUpdateDto cartItemRequestDto) {
```

File: src/main/java/com/bookappstore/controller/ShoppingCartController.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                cartItemRequestDto);
+    }
+
+    @DeleteMapping("/cart-items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Remove a book from the shopping cart",
+            description = "Removes a book from the user's shopping cart")
+    public ShoppingCartResponseDto deleteCartItem(
+            Authentication authentication,
+            @PathVariable Long cartItemId) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/ShoppingCartController.java]:
```suggestion
            @PathVariable @Positive Long cartItemId) {
```

File: src/main/java/com/bookappstore/mapper/CartItemMapper.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import org.mapstruct.MappingTarget;
+import org.mapstruct.Mappings;
+
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
+    @Mapping(target = "book", ignore = true)
+    CartItem toEntity(CartItemRequestDto cartItemRequestDto);
+
+    @Mapping(target = "book", ignore = true)
+    CartItem toEntity(CartItemUpdateDto cartItemUpdateDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/mapper/CartItemMapper.java]: Do you really need this mapper method?

File: src/main/java/com/bookappstore/model/CartItem.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Setter
+@Getter
+@SQLDelete(sql = "UPDATE cart_item SET is_deleted = true WHERE = id=?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "cart_item")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/CartItem.java]: Tables should be named in plural 
```suggestion
@Table(name = "cart_items")
```

File: src/main/java/com/bookappstore/model/CartItem.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/bookappstore/model/CartItem.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    @JoinColumn(name = "shopping_cart_id")
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/com/bookappstore/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Setter
+@Getter
+@SQLDelete(sql = "UPDATE shopping_cart SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "shopping_cart")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/ShoppingCart.java]:
```suggestion
@Table(name = "shopping_carts")
```
Check other tables

File: src/main/java/com/bookappstore/service/ShoppingCartService.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package com.bookappstore.service;
+
+import com.bookappstore.dto.cart.CartItemRequestDto;
+import com.bookappstore.dto.cart.CartItemUpdateDto;
+import com.bookappstore.dto.cart.ShoppingCartResponseDto;
+import org.springframework.transaction.annotation.Transactional;
+
+public interface ShoppingCartService {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/ShoppingCartService.java:9]:
```suggestion
```

File: src/main/java/com/bookappstore/service/ShoppingCartService.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    ShoppingCartResponseDto getShoppingCartDtoByUserId(Long id);
+
+    ShoppingCartResponseDto addCartItemByUserId(Long id,
+                                                CartItemRequestDto cartItemRequestDto);
+
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/ShoppingCartService.java]: ?

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+    @Override
+    public ShoppingCartResponseDto deleteCartItem(Long id, Long cartItemId) {
+        cartItemRepository.deleteById(cartItemId);
+        return shoppingCartMapper.toDto(getShoppingCartById(id));
+    }
+
+    public ShoppingCart getShoppingCartById(Long id) {
+        User user = userRepository.findById(id).orElseThrow(
+                () -> new RuntimeException("Can't find user by user id: " + id)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
                () -> new EntityNotFoundException("Can't find user by user id: " + id)
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        );
+        return shoppingCartRepository.findByUserId(user.getId()).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        return shoppingCartRepository.findByUserId(id).orElseThrow(
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+                () -> new EntityNotFoundException("Can't find shopping cart by id " + user.getId())
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
                () -> new EntityNotFoundException("Can't find shopping cart by id " + id)
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+    @Override
+    public ShoppingCartResponseDto addCartItemByUserId(
+            Long id,
+            CartItemRequestDto cartItemRequestDto) {
+        CartItem cartItem = cartItemMapper.toEntity(cartItemRequestDto);
+        ShoppingCart shoppingCart = getShoppingCartById(id);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(getShoppingCartById(id));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]: Don't need to call getShoppingCartById(id) twice

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public ShoppingCartResponseDto getShoppingCartDtoByUserId(Long id) {
+        ShoppingCart shoppingCartById = getShoppingCartById(id);
+        return shoppingCartMapper.toDto(shoppingCartById);
+    }
+
+    @Override
+    public ShoppingCartResponseDto addCartItemByUserId(
+            Long id,
+            CartItemRequestDto cartItemRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]: It's better to find book by Id, then shoppingCart by userId, 
cart.getCartItems().stream() - filter if cartItem already has this book by Id, ifPresentOrElse - change quality or addCartItemToCart(itemDto, book, cart))

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(getShoppingCartById(id));
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto updateCartItem(
+            Long id,
+            Long cartItemId,
+            CartItemUpdateDto cartItemUpdateDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]: Find ShoppingCart by userId first, then cartItemRepository.findByIdAndShoppingCartId

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+        CartItem cartItemDb = cartItemRepository.findById(cartItemId).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+                () -> new EntityNotFoundException("Can't find cart item by item id: " + cartItemId)
+        );
+        CartItem cartItemDto = cartItemMapper.toEntity(cartItemUpdateDto);
+        cartItemDb.setQuantity(cartItemDto.getQuantity());
+        cartItemRepository.save(cartItemDb);
+        return shoppingCartMapper.toDto(getShoppingCartById(id));
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java:74]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+    public ShoppingCartResponseDto deleteCartItem(Long id, Long cartItemId) {
+        cartItemRepository.deleteById(cartItemId);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```
ShoppingCart shoppingCart = repository.findByUserId(userId);
CartItem cartItem = itemRepository.findByIdAndShoppingCartId(....
shoppingCart.removeItemFromCart(cartItem);
```

where in ShoppingCart entity
```
public void removeItemFromCart(CartItem cartItem) {
        cartItems.remove(cartItem);
    }
```

File: src/main/java/com/bookappstore/service/impl/UserServiceImpl.java [hunk 24]
```diff
@@ -39,6 +42,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         Set<Role> defaultUserRoleSet = new HashSet<>();
         defaultUserRoleSet.add(userRole);
         user.setRoles(defaultUserRoleSet);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/UserServiceImpl.java]:
```suggestion
        user.setRoles(Set.of(userRole));
```

File: src/main/java/com/bookappstore/service/impl/UserServiceImpl.java [hunk 25]
```diff
@@ -39,6 +42,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
+
+        ShoppingCart shoppingCart = new ShoppingCart();
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/UserServiceImpl.java]:
```suggestion
        userRepo.save(user);
        cartService.createShoppingCart(user);
        return userMapper.toUserResponseDto(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/bookappstore/model/CartItem.java (hunk 11 above)
Comment (Elena-Bruyako) [reply src/main/java/com/bookappstore/model/CartItem.java]: not fixed

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+                },
+                () -> {
+                    CartItem newCartItem = cartItemMapper.toEntity(cartItemRequestDto);
+                    newCartItem.setBook(book);
+                    newCartItem.setShoppingCart(shoppingCart);
+                    cartItemRepository.save(newCartItem);
+                }
+        );
+
+        return shoppingCartMapper.toDto(getShoppingCartByUserId(userId));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        return shoppingCartMapper.toDto(shoppingCart);
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+    @Override
+    public ShoppingCartResponseDto addCartItem(
+            Long userId,
+            CartItemRequestDto cartItemRequestDto) {
+        Book book = bookRepository.findById(cartItemRequestDto.getBookId())
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find book")
+                );
+        ShoppingCart shoppingCart = getShoppingCartByUserId(userId);
+        Optional<CartItem> existingCartItem = shoppingCart.getCartItems().stream()
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        cart.getCartItems().stream()
                .filter(item -> item.getBook().getId().equals(itemDto.bookId()))
                .findFirst()
                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity() + itemDto.quantity()),
                        () -> addCartItemToCart(itemDto, book, cart));
        repository.save(cart);
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+        return shoppingCartMapper.toDto(shoppingCartById);
+    }
+
+    @Override
+    public ShoppingCartResponseDto addCartItem(
+            Long userId,
+            CartItemRequestDto cartItemRequestDto) {
+        Book book = bookRepository.findById(cartItemRequestDto.getBookId())
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find book")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]: Add id to message

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+        return shoppingCartMapper.toDto(getShoppingCartByUserId(userId));
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto updateCartItem(
+            Long id,
+            Long cartItemId,
+            CartItemUpdateDto cartItemUpdateDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart cart = repository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId()).orElseThrow(
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+                () -> new EntityNotFoundException("Can't find cart item by item id: " + cartItemId)
+        );
+        CartItem cartItemDto = cartItemMapper.toEntity(cartItemUpdateDto);
+        cartItem.setQuantity(cartItemDto.getQuantity());
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        cartItem.setQuantity(cartItemUpdateDto.getQuantity());
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto updateCartItem(
+            Long id,
+            Long cartItemId,
+            CartItemUpdateDto cartItemUpdateDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find cart item by item id: " + cartItemId)
+        );
+        CartItem cartItemDto = cartItemMapper.toEntity(cartItemUpdateDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
```

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+        cartItem.setQuantity(cartItemDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(getShoppingCartByUserId(id));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        return shoppingCartMapper.toDto(cart);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java (hunk 27 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/service/impl/ShoppingCartServiceImpl.java]: thanks, much better approach

**Review by Rommelua — CHANGES_REQUESTED** [review]
Not fixed previous comments

**Comment by VKhmil** [comment]
Perhaps I can use a different approach: splitting it into two services. One service would handle adding, deleting, and updating cart items, while another service, ShoppingCartService, would retrieve the entire cart and pass it to the controller. If necessary, the controller would then pass everything to CartItemService. I'm eager to hear your thoughts, dear mentors. Thank you.

File: src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("Can not find book by id: "
+                        + requestDto.getBookId())));
+
+        shoppingCart.getCartItems().add(cartItem);
+        shoppingCartRepository.save(shoppingCart);
+        return cartItemMapper.toDto(cartItem);
+    }
+
+    @Override
+    public CartItemResponseDto getById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java]: I didn't find where you use this method

File: src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+        return cartItemMapper.toDto(cartItemRepository.findById(id)
+                .orElseThrow(
+                        () -> new EntityNotFoundException(
+                                "Can not find cart item by id:" + id)));
+    }
+
+    @Override
+    public CartItemResponseDto delete(Long id) {
+        cartItemRepository.deleteById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java]: You need to find shopping cart by userId to clarify that you delete item in particular cart. Then find cartItem findByIdAndShoppingCartId and remove it from cartItems

File: src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java [hunk 35]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    public void addQuantity(Long userId, int quantity) {
+        shoppingCartRepository.findByUserId(userId).ifPresent(shoppingCart -> {
+            shoppingCart.getCartItems().forEach(cartItem ->
+                    cartItem.setQuantity(cartItem.getQuantity() + quantity));
+            shoppingCartRepository.save(shoppingCart);
+        });
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java]: About your comment with two services: 
first, we need to provide the business logic according to the task requirements. 
second, all methods in CartItemService provide the logic with shopping cart (add cart item, delete, change quantity in Shopping cart, adding cart item to shopping cart) is more appropriate for ShoppingCartService implementation.

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Left some comments

File: src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java (hunk 35 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/service/impl/CartItemServiceImpl.java]: Thank's more clear

**Comment by sarakhmen** [comment]
please pay attention to the previous mentor's comments. not all of them were taken into consideration and fixed

**Review by sarakhmen — CHANGES_REQUESTED** [review]
please pay attention to the previous mentor's comments. not all of them were taken into consideration and fixed

**Review by MAMentorFX — APPROVED** [review]

**Pull request merged**